### PR TITLE
#87 - Upload small content using put request when size is unknown

### PR DESCRIPTION
**Pull request by olegmoz**
Part of #87
Upload small content using put request in S3Storage even if size is unknown. Storage tries to read content up to some 10Mb, if content is bigger then multipart upload is used

### CODE REVIEW

**Comment by paulodamaso** [comment]
@g4s8 PLease assign me, I'm available

**Comment by 0crat** [comment]
This pull request [#130](https://github.com/artipie/asto/issues/130) is assigned to @vatavuk[/z](https://www.0crat.com/u/vatavuk), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/b3f3a8fe-2b75-4e1a-99af-4ec73116c2b8); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

File: src/main/java/com/artipie/asto/s3/S3Storage.java [hunk 1]
```diff
@@ -236,6 +235,61 @@ public S3Storage(final S3AsyncClient client, final String bucket) {
...
     }
 
+    /**
+     * Uploads content using put request.
+     *
+     * @param key Object key.
+     * @param content Object content to be uploaded.
+     * @return Completion stage which is completed when response received from S3.
+     */
+    private CompletableFuture<Void> put(final Key key, final Content content) {
```
Comment (Vatavuk) [inline src/main/java/com/artipie/asto/s3/S3Storage.java]: @olegmoz let's rename this method to `upload` to be consistent.

Comment (olegmoz) [reply src/main/java/com/artipie/asto/s3/S3Storage.java]: @Vatavuk sorry, but I think that current name is better. In S3 method for uploading file in single request is called `put`.

Comment (Vatavuk) [reply src/main/java/com/artipie/asto/s3/S3Storage.java]: @olegmoz ok

**Comment by olegmoz** [comment]
@g4s8 I cannot think of a way to do it right now, it will require to introduce some mockable abstraction inbetween S3Storage and S3Client. Not sure if it worth it, that’s a lot of work

**Comment by g4s8** [comment]
@olegmoz maybe some integration test with existing S3 client for tests then?

**Comment by olegmoz** [comment]
@g4s8 there is a test in S3StorageTest suite that uploads small value with unknown size, so this is tested. There is just no way to check what uploading method was used, it worked fine via multipart upload before this PR.

**Comment by g4s8** [comment]
> @g4s8 there is a test in S3StorageTest suite that uploads small value with unknown size, so this is tested. There is just no way to check what uploading method was used, it worked fine via multipart upload before this PR.

@olegmoz Ok, that's enough I think

**Comment by 0crat** [comment]
Job was finished in 21 hours, bonus for fast delivery is possible (see [§36](http://www.zerocracy.com/policy.html#36))

**Comment by 0crat** [comment]
Job audit: performer @vatavuk[/z](https://www.0crat.com/u/vatavuk)[/z](https://www.0crat.com/u/vatavuk) didn't make CR comments

**Pull request merged**